components/service-operator/internal/aws/cloudformation/client.go

```diff
@@ -0,0 +1,477 @@
+	case DeleteFailed, CreateFailed, RollbackFailed, UpdateRollbackFailed, RollbackComplete, UpdateRollbackComplete:
+		s.State = object.ErrorState
+	case DeleteInProgress, DeleteComplete:
+		s.State = object.DeletingState
+	case CreateComplete, UpdateComplete:
+		s.State = object.ReadyState
+	default:
+		s.State = object.ReconcilingState
+	}
+	// if object implements whitelisting of output keys, then update info
+	if w, ok := stack.(StackOutputWhitelister); ok {
+		if s.AWS.Info == nil {
+			s.AWS.Info = map[string]string{}
+		}
+		outputs, _ := r.Outputs(ctx, stack)
+		for _, whitelistedKey := range w.GetStackOutputWhitelist() {
+			s.AWS.Info[whitelistedKey] = outputs[whitelistedKey]
+		}
+	}
+	stack.SetStatus(s)
+}
+
+func (r *Client) events(ctx context.Context, stack Stack) ([]*StateEvent, error) {
+	eventsOutput, err := r.Client.DescribeStackEventsWithContext(ctx, &DescribeStackEventsInput{
+		StackName: aws.String(stack.GetStackName()),
+	})
+	if err != nil {
+		return nil, err
+	}
+	if eventsOutput == nil {
+		return []*StateEvent{}, nil
+	}
+	return eventsOutput.StackEvents, nil
+}
+
+// Exists checks if the stack has been provisioned
+func (r *Client) exists(ctx context.Context, stack Stack) (bool, error) {
+	_, err := r.get(ctx, stack)
+	if err == ErrStackNotFound {
+		return false, nil
+	} else if err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+func (r *Client) waitUntilCompleteState(ctx context.Context, stack Stack) (*State, error) {
+	return r.waitUntilState(ctx, stack, []string{
+		CreateComplete,
+		UpdateComplete,
+		UpdateRollbackComplete,
+		RollbackComplete,
+	})
+}
+
+func (r *Client) waitUntilDestroyedState(ctx context.Context, stack Stack) (*State, error) {
+	return r.waitUntilState(ctx, stack, []string{
+		DeleteComplete,
+	})
+}
+
+func (r *Client) waitUntilState(ctx context.Context, stack Stack, desiredStates []string) (*State, error) {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, context.DeadlineExceeded
+		default:
+			state, err := r.get(ctx, stack)
+			if IsNotFoundError(err) && in(DeleteComplete, desiredStates) {
+				// If we are waiting for DeleteComplete state and the
+				// stack has gone missing, consider this DeleteComplete
+				return &State{}, nil
+			} else if err != nil {
+				return nil, err
+			}
+			if in(*state.StackStatus, desiredStates) {
+				return state, nil
+			}
+		}
+		time.Sleep(r.PollingInterval)
+	}
+}
+
+func in(needle string, haystack []string) bool {
+	for _, s := range haystack {
+		if needle == s {
+			return true
+		}
+	}
+	return false
+}
+
+func IsNoUpdateError(err error) bool {
+	if err == nil {
+		return false
+	}
+	return strings.Contains(err.Error(), NoUpdatesErrMatch)
+}
+
+func IsNotFoundError(err error) bool {
+	if awsErr, ok := err.(awserr.Error); ok {
+		if awsErr.Code() == "ResourceNotFoundException" {
+			return true
+		} else if awsErr.Code() == "ValidationError" && strings.Contains(awsErr.Message(), NoExistErrMatch) {
+			return true
+		}
+	}
```

Comment:
I kind of wonder if we should say that if it's not an awserror.Error then return false?

Reply:
What I said makes literally no sense and I no longer follow myself.